File: packages/acl/dart/lib/src/token.dart
final _segmentPattern = RegExp(r"^[a-z][a-z0-9_]*$");

/// Grammar-level parts of an ACL token: `{domain}:{action}[:{qualifier}]`.
final class TokenParts {
  const TokenParts({required this.domain, required this.action, this.qualifier});

  /// The permission domain, e.g. `post`.
  final String domain;

  /// The action within [domain], e.g. `create`.
  final String action;

  /// The optional qualifier narrowing [action], e.g. `own`.
  final String? qualifier;

  /// Encodes these parts back into their string form.
  String encode() => qualifier == null ? "$domain:$action" : "$domain:$action:$qualifier";
}

/// Exception thrown when a string does not satisfy the ACL token grammar.
final class AclTokenFormatException implements Exception {
  const AclTokenFormatException(this.token, this.reason);

  /// The offending token.
  final String token;

  /// Why the token is malformed.
  final String reason;

  @override
  String toString() => 'Invalid ACL token "$token": $reason';
}

/// Parses [token] into its grammar-level parts.
///
/// This validates the token grammar only; whether the domain, action, or
/// qualifier actually exist is a schema-level concern (see the generated
/// `parseAclToken`). Throws [AclTokenFormatException] on malformed input.
TokenParts parseToken(String token) {
  final segments = token.split(":");
  if (segments.length < 2 || segments.length > 3) {
    throw AclTokenFormatException(token, "expected 2 or 3 segments, found ${segments.length}");
  }
  for (final segment in segments) {
    if (!_segmentPattern.hasMatch(segment)) {
      throw AclTokenFormatException(token, 'invalid segment "$segment"');
    }
  }
  return TokenParts(
    domain: segments[0],
    action: segments[1],
    qualifier: segments.length == 3 ? segments[2] : null,
  );
}

File: packages/acl/ts/src/token.ts
const SEGMENT_PATTERN = /^[a-z][a-z0-9_]*$/;

/** Grammar-level parts of an ACL token: `{domain}:{action}[:{qualifier}]`. */
export interface TokenParts {
    domain: string;
    action: string;
    qualifier?: string;
}

/** Error thrown when a string does not satisfy the ACL token grammar. */
export class AclTokenError extends Error {
    readonly token: string;

    constructor(token: string, reason: string) {
        super(`Invalid ACL token "${token}": ${reason}`);
        this.name = "AclTokenError";
        this.token = token;
    }
}

/**
 * Parses `token` into its grammar-level parts.
 *
 * This validates the token grammar only; whether the domain, action, or
 * qualifier actually exist is a schema-level concern (see the generated
 * `isAclToken` guard). Throws {@link AclTokenError} on malformed input.
 */
export function parseToken(token: string): TokenParts {
    const segments = token.split(":");
    if (segments.length < 2 || segments.length > 3) {
        throw new AclTokenError(token, `expected 2 or 3 segments, found ${segments.length}`);
    }
    for (const segment of segments) {
        if (!SEGMENT_PATTERN.test(segment)) {
            throw new AclTokenError(token, `invalid segment "${segment}"`);
        }
    }
    const [domain, action, qualifier] = segments;
    return { domain, action, qualifier };
}

/** Encodes token parts back into their `{domain}:{action}[:{qualifier}]` string form. */
export function formatToken(parts: TokenParts): string {
    return parts.qualifier === undefined
        ? `${parts.domain}:${parts.action}`
        : `${parts.domain}:${parts.action}:${parts.qualifier}`;
}
